enhancements/automated-resource-request-scaling.md

```diff
@@ -0,0 +1,855 @@
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+If applicable, how will the component be upgraded and downgraded? Make sure this
+is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade in order to keep previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade in order to make use of the enhancement?
+
+Upgrade expectations:
+- Each component should remain available for user requests and
+  workloads during upgrades. Ensure the components leverage best practices in handling [voluntary disruption](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/). Any exception to this should be
+  identified and discussed here.
+- Micro version upgrades - users should be able to skip forward versions within a
+  minor release stream without being required to pass through intermediate
+  versions - i.e. `x.y.N->x.y.N+2` should work without requiring `x.y.N->x.y.N+1`
+  as an intermediate step.
+- Minor version upgrades - you only need to support `x.N->x.N+1` upgrade
+  steps. So, for example, it is acceptable to require a user running 4.3 to
+  upgrade to 4.5 with a `4.3->4.4` step followed by a `4.4->4.5` step.
+- While an upgrade is in progress, new component versions should
+  continue to operate correctly in concert with older component
+  versions (aka "version skew"). For example, if a node is down, and
+  an operator is rolling out a daemonset, the old and new daemonset
+  pods must continue to work correctly even while the cluster remains
+  in this partially upgraded state for some time.
+
+Downgrade expectations:
+- If an `N->N+1` upgrade fails mid-way through, or if the `N+1` cluster is
+  misbehaving, it should be possible for the user to rollback to `N`. It is
+  acceptable to require some documented manual steps in order to fully restore
+  the downgraded cluster to its previous state. Examples of acceptable steps
+  include:
+  - Deleting any CVO-managed resources added by the new version. The
+    CVO does not currently delete resources that no longer exist in
+    the target version.
+
+### Version Skew Strategy
+
+Clusters deployed from versions that do not support scaling will show
+a value of `0`, causing operators that support scaling to use their
+default settings.
+
+Operators from versions that do not support scaling will ignore the
+new field until they are upgraded.
+
+The value of `DefaultControlPlaneCPUCount` constant may change between
+releases. The value from the library will be vendored into each
+operator, so the scaling formula for a given version of an operator
+will be correct at all times during an upgrade, even if different
+operators are using different base values.
```

Comment:
I don't think this is quite right.  Based on your goal:

> The current resource requirements for an OpenShift control plane exceed the size desired by some customers running on bare metal.

where you have *requirements*, not *requests*, and [this comment][1], the current proposal would result in:

1. Update initiated.
2. New CVO comes in, with new `DefaultControlPlaneCPUCount`.  Say it's lower.  CVO starts working through its [manifest graph][2].
3. CVO gets to operator A's Deployment, and it updates to match the new manifest, with the scaled request value.
4. New operator A pod comes up with its new `DefaultControlPlaneCPUCount`.  It fetches the Infrastructure object, sees a low `controlPlaneCPULimit`, and says "ooh, I'm really going to have to reign it in", and triples some poll periods and adds some sleeps or whatever.

Having the `request` reduction in (3) match the actual-consumption reduction from (4) means that we really want `DefaultControlPlaneCPUCount` to be uniform throughout the cluster.  That's a lot of components that will need to bump vendored libraries to pick this up; I'm not convinced we won't screw it up ;).

And more broadly, there is going to be some resource consumption drop in most processes just because their single-node workload is lighter than their Standard Cluster workload.  But this proposal is not asking the process to figure out single-node-ness.  If the process throttles back polling periods and whatnot by the amount needed to take its Standard Cluster CPU consumption down by the requested amount, and then we run it with single-node workloads, it may end up starving itself more than it needs to, resulting in even higher latencies, etc. than necessary.  Without something actually looking at measured consumption in order to tune whatever the in-process knobs end up being ("ah, I'm over my `requests` more often than under them in this past second, better reign it in"), it seems like it is going to be hard to reliably approach the `controlPlaneCPULimit` target.

[1]: https://github.com/openshift/enhancements/pull/628#discussion_r570578301
[2]: https://github.com/openshift/cluster-version-operator/blob/dfc8494a4ed9b4e68d7511ae989cbf3225d65e65/docs/user/reconciliation.md#manifest-graph

Reply:
I thought a `const` in a library would expose fewer implementation details, but your counterargument is good. If we exposed `DefaultControlPlaneCPUCount` via the Infrastructure API, what would be responsible for updating it during an upgrade? Do we have anything that writes to that API? Should we use a different resource, so we can ship a static manifest with the value and have the CVO apply it as part of the upgrade?

@browsell correct me if I'm wrong, but I thought this was mostly about being able to schedule the workloads, and that they would be pinned to CPUs to throttle them without individual components having to constantly adjust their resource consumption.

As far as scaling the operators requests, I had been expecting the work done to reduce the resource constraints on operators deployed by the CVO to be enough that the CVO would not need to take further action, so the API would only be used by those operators but not "on" them. Maybe that's not a safe assumption, though.

Reply:
Target application for this deployment model would mostly use guaranteed/exclusive cpus. The expectation for the remaining applications pods (i.e. burstable) is that they would be tightly engineered to optimize the cpu footprint. 

Reply:
> Target application for this deployment model would mostly use guaranteed/exclusive cpus. The expectation for the remaining applications pods (i.e. burstable) is that they would be tightly engineered to optimize the cpu footprint.

When you say "application pods" do you mean only user-provided workloads that are not part of OpenShift or other add-ons that make up "the platform"?

Reply:
User provided workloads that are not part of openshift

Reply:
OK, good. So, I think @wking's question applies to things that are part of OpenShift, and how we ensure that those actually fit within the number of CPUs allowed. We've been talking about scaling the requests, but the concern is about components that use more CPU than they request, at least some of the time. Do we anticipate the scaling behavior, in whatever implementation, to provide enough overhead to support those spikes in use without excessive throttling leading to latency issues?

Reply:
On average the actual usage of openshift pods would be less than the targeted number of allowed cpu's. There are a small number of spikes that would exceed it. We should consider providing some overhead for spikes but at the same time I don't think we should engineer for the worst case spike, rather we should look at the cause and address it.